include/GafferImage/FlatImageSource.h

```diff
@@ -0,0 +1,82 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2019, Image Engine Design Inc. All rights reserved.
+//  Copyright (c) 2015, Nvizible Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERIMAGE_FLATIMAGESOURCE_H
+#define GAFFERIMAGE_FLATIMAGESOURCE_H
+
+#include "GafferImage/ImageNode.h"
+
+namespace GafferImage
+{
+
+/// The FlatImageSource is a base class for several nodes which produce flat
+/// images.  It takes care of returning flat values for the plugs used by
+/// deep images.
+class GAFFERIMAGE_API FlatImageSource : public ImageNode
```

Comment:
If you think it's reasonable to rip it out, I would be happy to do that.

I've been starting to consider the entirety of IECore::ImagePrimitive as heading on it's way to deprecation - it doesn't seem like there's much reason to use that structure instead of Gaffer or OpenImageIO.  It may not be fully deprecated any time soon, but I would consider ripping out ObjectToImage a step in the right direction.

Reply:
> I've been starting to consider the entirety of IECore::ImagePrimitive as heading on it's way to deprecation 

Me too. I think the main hurdle there is our dependence on it in IECoreGL. I don't expect us to get there all that soon, but anything we can do to reduce dependence along the way is worth it.

> I would consider ripping out ObjectToImage a step in the right direction.

Definitely. I'm keenly aware that we're on the brink of having this merged though, so I'm reluctant to send you down a potential rabbit hole. How about we save it as a followup, ideally immediately after merging?